File: src/tree_parity_machine/neuron/HiddenNeuron.java
package tree_parity_machine.neuron;

import learning_algorithm.LearningParadigm;
import tree_parity_machine.NeuralNetException;
import utils.Random;

public class HiddenNeuron extends Neuron {

    public HiddenNeuron(int inputs, int leftBound, int rightBound, LearningParadigm paradigm) {
        this.inputs = inputs;
        this.paradigm = paradigm;
        weights = new double[inputs];
        Random.setBounds(leftBound, rightBound);
    }

    public void init() {
        for (int i = 0; i < inputs; i++)
            weights[i] = Random.getInt();
    }

    @Override
    public void changeWeights(double[] input, int outputTPM, LearningParadigm paradigm) {
        for (int i = 0; i < input.length; i++)
            switch (paradigm) {
                case HEBIAN:
                    weights[i] += input[i] * outputTPM;
                    break;
                case ANTI_HEBBIAN:
                    weights[i] -= input[i] * outputTPM;
                    break;
                case RANDOM_WALK:
                    weights[i] += input[i];
                    break;
            }
    }

    public int getOutput(double[] input) throws NeuralNetException {
        if (input == null || input.length != inputs)
            throw new NeuralNetException("Входной вектор не соответствует кол-ву весовых коэффициентов");
        double sum = 0;
        for (int i = 0; i < inputs; i++)
            sum += weights[i] * input[i];
        output = sum > 0 ? 1 : -1;
        return output;
    }

}

File: src/main/kotlin/com/jvvladimir/machine/tree_machine/neuron/HiddenNeuron.kt
package com.jvvladimir.machine.tree_machine.neuron

import com.jvvladimir.machine.learning.LearningParadigm
import com.jvvladimir.machine.tree_machine.NeuralNetException
import com.jvvladimir.machine.utils.MyRandom.getInt
import com.jvvladimir.machine.utils.MyRandom.getInts

class HiddenNeuron(inputs: Int, leftBound: Int, rightBound: Int, paradigm: LearningParadigm) : Neuron() {

    init {
        this.inputs = inputs
        this.paradigm = paradigm
        weights = IntArray(inputs)
        this.leftBound = leftBound
        this.rightBound = rightBound
    }

    override fun init() {
        weights = getInts(inputs, leftBound, rightBound)
    }

    override fun changeWeights(input: IntArray, outputTPM: Int) {
        for (i in input.indices) {
            val dW = input[i] * outputTPM
            weights[i] = when (paradigm) {
                LearningParadigm.HEBBIAN -> weights[i] + dW
                LearningParadigm.ANTI_HEBBIAN -> weights[i] - dW
                LearningParadigm.RANDOM_WALK -> weights[i] + input[i]
            }
            if (weights[i] > rightBound) weights[i] = rightBound else if (weights[i] < leftBound) weights[i] = leftBound
        }
    }

    override fun getOutput(input: IntArray): Int {
        if (input.size != inputs) throw NeuralNetException("Входной вектор не соответствует кол-ву весовых коэффициентов")
        var sum = 0
        for (i in 0 until inputs) sum += weights[i] * input[i]
        output = if (sum > 0) 1 else -1
        return output
    }
}